upup/models/cloudup/resources/addons/karpenter.sh/k8s-1.19.yaml.template

```diff
@@ -0,0 +1,677 @@
+metadata:
+  name: defaulting.webhook.provisioners.karpenter.sh
+webhooks:
+- admissionReviewVersions: ["v1"]
+  clientConfig:
+    service:
+      name: karpenter-webhook
+      namespace: 'kube-system'
+  failurePolicy: Fail
+  sideEffects: None
+  name: defaulting.webhook.provisioners.karpenter.sh
+  rules:
+  - apiGroups:
+    - karpenter.sh
+    apiVersions:
+    - v1alpha5
+    resources:
+    - provisioners
+      provisioners/status
+    operations:
+    - CREATE
+    - UPDATE
+---
+# Source: karpenter/templates/webhook/webhooks.yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: validation.webhook.provisioners.karpenter.sh
+webhooks:
+- admissionReviewVersions: ["v1"]
+  clientConfig:
+    service:
+      name: karpenter-webhook
+      namespace: 'kube-system'
+  failurePolicy: Fail
+  sideEffects: None
+  name: validation.webhook.provisioners.karpenter.sh
+  rules:
+  - apiGroups:
+    - karpenter.sh
+    apiVersions:
+    - v1alpha5
+    resources:
+    - provisioners
+      provisioners/status
+    operations:
+    - CREATE
+    - UPDATE
+---
+# Source: karpenter/templates/webhook/webhooks.yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: validation.webhook.config.karpenter.sh
+webhooks:
+- admissionReviewVersions: ["v1"]
+  clientConfig:
+    service:
+      name: karpenter-webhook
+      namespace: 'kube-system'
+  failurePolicy: Fail
+  sideEffects: None
+  name: validation.webhook.config.karpenter.sh
+  objectSelector:
+    matchLabels:
+      app.kubernetes.io/part-of: karpenter
+
+{{ range $name, $spec := GetNodeInstanceGroups }}
+{{ if eq $spec.InstanceManager "Karpenter" }}
+---
+apiVersion: karpenter.sh/v1alpha5
+kind: Provisioner
+metadata:
+  name: {{ $name }}
+spec:
+  requirements:
+    - key: karpenter.sh/capacity-type
+      operator: In
+      values: ["spot"]
+    - key: kubernetes.io/arch
+      operator: In
+      values: ["{{ SupportedArchitecture $spec.Image }}"]
+    - key: "node.kubernetes.io/instance-type"
+      operator: In
+      values:
+      - {{ $spec.MachineType }}
+      {{ with $spec.MixedInstancesPolicy }}
+      {{ range $key := .Instances }}
+      - {{ $key }}
+      {{ end }}
+      {{ end }}
+{{ with $spec.Taints }}
+  taints:
+    {{ range $taintString := $spec.Taints }}
+    {{ $taint := ParseTaint $taintString }}
+  - key: {{ $taint.key }}
+    effect: {{ $taint.effect }}
+    {{ if $taint.value }}
+    value: {{ $taint.value }}
+    {{ end }}
+{{ end }}
+{{ end }}
+{{ with $spec.NodeLabels }}
+  labels:
+  {{ range $key, $value := . }}
+    {{ $key }}: {{ $value }}
+  {{ end }}
+{{ end }}
+  limits:
+    resources:
+      cpu: 1000
+  provider:
+    instanceProfile: nodes.{{ ClusterName }}
+    launchTemplate: {{ $name }}
+    securityGroupSelector:
+      Name: nodes.{{ ClusterName }}
+    subnetSelector:
+      kops.k8s.io/karpenter/{{ $name }}: "*"
```

Comment:
There's a todo to support subnets?

Reply:
We do. The code iterates over the mentioned subnets and add this tag to it. It may be that this doesn't work with shared subnets. But that's the intention.
Ideally, Karpenter would allow for multiple selectors, but currently they don't.

Reply:
We turn off `cluster.spec.tagSubnets` and manage the subnets with Terraform. It would be a real pain to have to tag them for each IG, especially since the Terraform code/config doesn't otherwise need to be changed much.

Possibly this should be feedback for upstream.

In our use cases, we have some IGs that spread across all 3 AZs and some that only go into one AZ, "a". So allowing a configurable subnetSelector would work well for us.

Reply:
Should be easy enough to make subnetSelector configurable, but usability-wise having multiple selectors would be better.

I also don't think one can directly specify subnet IDs. Our `Find` do not require subnet name in spec to match the Name tag if ID is specified, so if there is a mismatch there, using a tag-based selector would fail.

I suggest at minimum having an upstream issue for multiple selectors.